elara/elarautil.py

```diff
@@ -4,62 +4,108 @@
 
 This source code is licensed under the BSD-style license found in the LICENSE file in the root directory of this source tree.
 """
+from typing import Dict
 
-from cryptography.fernet import Fernet
-import json
+import msgpack
 import os
-from .exceptions import FileAccessError, FileKeyError
+from zlib import crc32
+
+from cryptography.fernet import Fernet
+
+from .exceptions import FileAccessError, FileKeyError, LoadChecksumError, LoadIncompatibleDB
 
 
 class Util:
     @staticmethod
-    def readJSON(obj):
-        try:
-            curr_db = json.load(open(obj.path, "rb"))
-        except Exception:
-            # print("Read JSON error. File might be encrypted. Run in secure mode.")
-            raise FileAccessError(
-                "Read JSON error. File might be encrypted. Run in secure mode with key path."
-            )
-        return curr_db
+    def check_mag(mag):
+        return mag == b"ELDB"
 
     @staticmethod
-    def storeJSON(obj):
-        try:
-            json.dump(obj.db, open(obj.path, "wt"), indent=4)
-        except Exception:
-            raise FileAccessError(
-                "Store JSON error. File might be encrypted. Run in secure mode with key path."
-            )
+    def check_encrypted(version):
+        # if msb of version number is set the db is encrypted
+        return (version & (1 << 15)) != 0
```

Comment:
Can you ellaborate (1 << 15)

 `buffer += (obj.db_format_version | 1 << 15).to_bytes(2, "little")` here as well

Reply:
Just want to understand the versioning better

Reply:
it's just a way to flag if a file is using encryption(you could add more stuff like compression down the line) by reading it's first few bytes, the msb is the leftmost bit on this 16 bit number so if I shift 1 by 15 I get a mask that I can use there to test this condition

Reply:
I guess it could be replaced by a constant but I guess me being used to working with shifts and masking in C reached for this first

Reply:
the constant would be 0x8000 or 32768

Reply:
> Just want to understand the versioning better

The idea is if the format gets any more breaking changes elara will simply refuse to load the db file to avoid data corruption on load

Reply:
Okay, I kind of understood what you mean but could you maybe share a link to an article (or any such resource) covering this.

Reply:
about which part more specifically the bit masking and flags or the versioning part?

Reply:
bit masking and flags, but do send both if you can 😛

Reply:
Versioning there is not much to it, it's what I described, flags and bit twiddling on the other hand can be a deep rabbit hole, here you go:
* https://hackmykaam.com/bit-masking-cheat-sheet/
* https://en.wikipedia.org/wiki/Mask_(computing)
* https://www.arduino.cc/en/Tutorial/Foundations/BitMask
* https://www.interviewcake.com/concept/java/bit-shift

another example is say I want to set the 50th bit of a 64 bit int to 1 I can use a large constant or express it as `1 << 49`
these links should have you covered, btw the constant to my example is 562949953421312 in decimal or 0x2000000000000 hex.

Reply:
I hope my example with the very large constants shows how using shifts is generally more expressive